text/0000-trait-alias.md

```diff
@@ -0,0 +1,266 @@
+3. The `where` clause, if present, only contains bounds on `Self`.
+
+Some examples:
+
+```rust
+trait Sink = Sync;
+trait ShareableIterator = Iterator + Sync;
+trait PrintableIterator = Iterator<Item=i32> + Display;
+trait IntIterator = Iterator<Item=i32>;
+
+fn foo1<T: ShareableIterator>(...) { ... } // ok
+fn foo2<T: ShareableIterator<Item=i32>>(...) { ... } // ok
+fn bar1(x: Box<ShareableIterator>) { ... } // ERROR: associated type not specified
+fn bar2(x: Box<ShareableIterator<Item=i32>>) { ... } // ok
+fn bar3(x: Box<PrintableIterator>) { ... } // ERROR: too many traits (*)
+fn bar4(x: Box<IntIterator + Sink + 'static>) { ... } // ok (*)
+```
+
+The lines marked with `(*)` assume that [#24010](https://github.com/rust-lang/rust/issues/24010) is fixed.
+
+# Teaching
+[teaching]: #teaching
+
+[Traits](https://doc.rust-lang.org/book/traits.html) are obviously a huge prerequisite. Traits aliases could be introduced at the end of
+that chapter.
+
+Conceptually, a *trait alias* is a syntax shortcut used to reason about one or more trait(s). Inherently, the *trait alias* is usable
+in a limited set of places:
+
+- as a *bound*: exactly like a *trait*, a *trait alias* can be used to constraint a type (type parameters list, where-clause)
+- as a *trait object*: same thing as with a *trait*, a *trait alias* can be used as a *trait object* if it fits object safety restrictions (see above in the [semantics](#semantics) section)
+- in an [`impl Trait`](https://github.com/rust-lang/rfcs/blob/master/text/1522-conservative-impl-trait.md)
+
+Examples should be showed for all of the three cases above:
+
+#### As a bound
+
+```rust
+trait StringIterator = Iterator<Item=String>;
+
+fn iterate<SI>(si: SI) where SI: StringIterator {} // used as bound
+```
+
+#### As a trait object
+
+```rust
+fn iterate_object(si: &StringIterator) {} // used as trait object
+```
+
+#### In an `impl Trait`
+
+```rust
+fn string_iterator_debug() -> impl Debug + StringIterator {} // used in an impl Trait
+```
+
+As shown above, a *trait alias* can substitute associated types. It doesn’t have to substitute them all. In that case, the *trait alias*
+is left incomplete and you have to pass it the associated types that are left. Example with the
+[tokio case]([tokio example](#second-example-ergonomic-collections-and-scrapping-boilerplate):
+
+```rust
+pub trait Service {
+  type Request;
+  type Response;
+  type Error;
+  type Future: Future<Item=Self::Response, Error=Self::Error>;
+  fn call(&self, req: Self::Request) -> Self::Future;
+}
+
+trait HttpService = Service<Request = http::Request, Response = http::Response, Error = http::Error>;
+
+trait MyHttpService = HttpService<Future = MyFuture>; // assume MyFuture exists and fulfills the rules to be used in here
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Adds another construct to the language.
+
+- The syntax `trait TraitAlias = Trait` requires lookahead in the parser to disambiguate a trait from a trait alias.
```

Comment:
Ultimately, this will depend on what kind of "parser generation" strategy you are using, but there isn't an inherent need for lookahead here. The key point is that the same syntax one can use on a trait, one can use on a trait alias, right up until you find either an `=` (trait alias) or a `where/{` (trait item). The problem would be if there were overlapping but not identical parts in the prefix.